src/ScheduleFunctions.cpp

```diff
@@ -1451,30 +1451,133 @@ class InjectFunctionRealization : public IRMutator {
+                    adj_list[func_name_to_index[level.func()]][level.stage_index()].push_back({(int)i, 0});
+                }
+            }
+            for (size_t j = 0; j < funcs[i].updates().size(); ++j) {
+                const auto &level = funcs[i].updates()[j].schedule().fuse_level().level;
+                if (!level.is_root() && !level.is_inlined()) {
+                    stage_dependencies[i][j + 1]++;
+                    adj_list[func_name_to_index[level.func()]][level.stage_index()].push_back({(int)i, (int)j + 1});
+
+                    // Let say that we have a stage f.update(p), which is scheduled to be computed_with
+                    // another stage g.update(q) (like f.update(p).compute_with(g.update(q), var)).
+                    // This means that the loop for f.update(p) will be injected into the loop for g.update(q).
+                    // Given that, for this to be correct, all stages of f up until (p - 1) must come
+                    // before g.update(q).
+                    // However, there is a special case here when two or more consecutive stages are computed
+                    // with the same stage. In this case, we won't be adding back edge to avoid creating cyclic
+                    // dependency.
+                    if (!(prev_level.func() == level.func() && prev_level.stage_index() == level.stage_index())) {
+                        for (size_t k = 0; k < j + 1; k++) {
+                            stage_dependencies[func_name_to_index[level.func()]][level.stage_index()]++;
+                            adj_list[i][k].push_back({func_name_to_index[level.func()], level.stage_index()});
+                        }
+                    }
+                    prev_level = level;
+                }
+            }
+        }
+
+        size_t complete_count = 0;
+        vector<size_t> stage_index(funcs.size());
+        // This basically computes topological order, but exploits the fact that stages of the function
+        // form linear order. Basically, we have a set of indices that point to the current stages
+        // for each of the functions and should be considered as a next stage in the general order. They
+        // are added to the order, only if all of their dependencies have been added already.
+        while (complete_count < funcs.size()) {
+            bool progress_made = false;
+            for (size_t i = 0; i < funcs.size(); i++) {
+                // We already added all stages of this function, so proceed to the next function.
+                if (stage_index[i] == stage_dependencies[i].size()) {
+                    continue;
+                }
+                // Proceed as far as we can, so stages of the same function are bundled together.
+                while (stage_index[i] < stage_dependencies[i].size()) {
+                    if (stage_dependencies[i][stage_index[i]] > 0) {
+                        break;
+                    }
+                    progress_made = true;
```

Comment:
Super minor nit: I think this should be at the end of the loop, after the progress has been made :)

Reply:
Fair point, updated.